### PR TITLE
Impose include/src/test code management for c libgpuarray

### PR DESCRIPTION
**Pull request by tsirif**
- Fix header guards to comply with google style
- Add comments on else and endif preprocessor clauses
- Change angled includes to quoted for project's headers

These things were discussed on #193. What do you think @abergeron?

### CODE REVIEW

File: gpuarray/config/gen_types.py [hunk 1]
```diff
@@ -45,9 +46,9 @@ def add_type(name, C, sz):
 typedef unsigned __int32 uint32_t;
 typedef signed __int64 int64_t;
 typedef unsigned __int64 uint64_t;
-#else
+#else  // _MSC_VER
```
Comment (abergeron) [inline gpuarray/config/gen_types.py]: This is a C++ style comment which is not supported in C89 (MSVC).

File: gpuarray/include/cache.h [hunk 2]
```diff
@@ -1,8 +1,9 @@
-#ifndef CACHE_H
-#define CACHE_H
+#ifndef LIBGPUARRAY_CACHE_H
```
Comment (abergeron) [inline gpuarray/include/cache.h]: Why move this in include?  It is not a public header.

File: gpuarray/config/private_config.h.in [hunk 3]
```diff
@@ -1,54 +1,53 @@
 /******************************************************************
```
Comment (abergeron) [inline gpuarray/config/private_config.h.in]: The new name just makes it more annoying to include.

File: gpuarray/include/gpuarray/blas.h [hunk 4]
```diff
@@ -1,12 +1,12 @@
-#ifndef GPUARRAY_BLAS_H
-#define GPUARRAY_BLAS_H
+#ifndef LIBGPUARRAY_GPUARRAY_BLAS_H
+#define LIBGPUARRAY_GPUARRAY_BLAS_H
 
-#include <gpuarray/buffer_blas.h>
-#include <gpuarray/array.h>
```
Comment (abergeron) [inline gpuarray/include/gpuarray/blas.h]: The includes inside the .h should really keep the angle brackets since they will be used inside user projects.

File: gpuarray/include/gpuarray/buffer_blas.h [hunk 5]
```diff
@@ -1,12 +1,12 @@
-#ifndef GPUARRAY_BUFFER_BLAS_H
-#define GPUARRAY_BUFFER_BLAS_H
+#ifndef LIBGPUARRAY_GPUARRAY_BUFFER_BLAS_H
+#define LIBGPUARRAY_GPUARRAY_BUFFER_BLAS_H
```
Comment (abergeron) [inline gpuarray/include/gpuarray/buffer_blas.h]: Is there any benefit to LIBGPUARRAY_GPUARRAY instead of GPUARRAY?

File: gpuarray/include/private.h [hunk 6]
```diff
@@ -1,30 +1,29 @@
-#ifndef _PRIVATE
-#define _PRIVATE
+#ifndef LIBGPUARRAY_PRIVATE_H
+#define LIBGPUARRAY_PRIVATE_H
```
Comment (abergeron) [inline gpuarray/include/private.h]: Again this is a private header.  Why is it in include?

**Comment by abergeron** [comment]
I have a bunch of comments in a not-so-ordered fashion:

1) Point 1 what to leave the angle bracket in the public headers.  This is not done here.
2) Also all the files have moved into sometimes much more awkward names, with useless repetition.  I personally dislike projects where there are fives levels of folders until you get to the source.
3) Adding a include subdirectory is not a bad idea, but only the public headers should go in there (those that are currently in src/gpuarray/)  All the others should stay with the source.
4) In some instance this feels like change for change's sake.
5) Spending two days on this is a bit too much for GSOC, so you should probably move on to platoon.
6) It currently needs a rebase.  But I would strongly advise to move on to platoon first and come back to this once you finish that.  The goal of the project is to have multi-gpu reductions.

File: gpuarray/config/gen_types.py (hunk 1 above)
Comment (tsirif) [inline gpuarray/config/gen_types.py]: Is this specific for `#else` or for `#endif` too? What style is advised in general?

File: gpuarray/include/gpuarray/buffer_blas.h (hunk 5 above)
Comment (tsirif) [inline gpuarray/include/gpuarray/buffer_blas.h]: Google style advises <project_name>_<directories>_<file name> for a header guard. I also feel that this is redundant here, but prefixing LIBGPUARRAY to other guards (e.g. cache.h) is beneficial I think.

File: gpuarray/config/private_config.h.in (hunk 3 above)
Comment (tsirif) [inline gpuarray/config/private_config.h.in]: Should I leave it in src then?

**Comment by tsirif** [comment]
Yes this is at least secondary. I will catch up.

**Comment by tsirif** [comment]
I changed things according to what you noted. Except for the comment style on `#else` clause, should I remove them?

File: gpuarray/include/gpuarray/buffer_blas.h (hunk 5 above)
Comment (abergeron) [inline gpuarray/include/gpuarray/buffer_blas.h]: Yes, a header guard is basically mandatory.  I'm just arguing about the tag itself.

File: .gitignore [hunk 7]
```diff
@@ -21,5 +21,4 @@ distribute*tar.gz
 *.log
 pygpu/*.c
 pygpu/*.h
-src/private_config.h
```
Comment (abergeron) [inline .gitignore]: Why was this removed?

Comment (abergeron) [inline .gitignore]: I see you've made a gpuarray/.gitignore file.  This is ok.

File: gpuarray/CMakeLists.txt [hunk 8]
```diff
@@ -107,11 +92,11 @@ if (CUDA_FOUND)
     add_definitions(-DNVCC_BIN=${CUDA_NVCC_EXECUTABLE})
   endif()
 
-  list(APPEND _GPUARRAY_SRC gpuarray_buffer_cuda.c)
+  list(APPEND GPUARRAY_SRC src/gpuarray/cuda/buffer_cuda.c)
   add_definitions(-DWITH_CUDA)
   include_directories(${CUDADRV_INCLUDE})
 
-  list(APPEND _GPUARRAY_SRC gpuarray_blas_cuda_cublas.c)
+  list(APPEND GPUARRAY_SRC src/gpuarray/cuda/blas_cuda_cublas.c)
```
Comment (abergeron) [inline gpuarray/CMakeLists.txt]: This could be named blas_cublas.c

File: gpuarray/CMakeLists.txt [hunk 9]
```diff
@@ -107,11 +92,11 @@ if (CUDA_FOUND)
     add_definitions(-DNVCC_BIN=${CUDA_NVCC_EXECUTABLE})
   endif()
 
-  list(APPEND _GPUARRAY_SRC gpuarray_buffer_cuda.c)
+  list(APPEND GPUARRAY_SRC src/gpuarray/cuda/buffer_cuda.c)
```
Comment (abergeron) [inline gpuarray/CMakeLists.txt]: This can probably be named just buffer.c

File: gpuarray/CMakeLists.txt [hunk 10]
```diff
@@ -123,36 +108,30 @@ if (CUDA_FOUND)
 
   if(NCCL_FOUND)
     message(STATUS "Building with NCCL")
-    set(BUILD_WITH_COLLECTIVES 1 PARENT_SCOPE)
     add_definitions(-DWITH_CUDA_NCCL)
-    list(APPEND _GPUARRAY_SRC gpuarray_collectives_cuda_nccl.c)
+    list(APPEND GPUARRAY_SRC src/gpuarray/cuda/collectives_cuda_nccl.c)
```
Comment (abergeron) [inline gpuarray/CMakeLists.txt]: collectives_nccl.c

File: gpuarray/src/private_config.h [hunk 11]
```diff
@@ -0,0 +1,53 @@
+/******************************************************************
```
Comment (abergeron) [inline gpuarray/src/private_config.h]: This file shouldn't be in the repo.

Comment (tsirif) [inline gpuarray/src/private_config.h]: I forgot to update .gitgnore while moving it. I rebase interactively. What should be done with the comments in #else?

File: gpuarray/test/CMakeLists.txt [hunk 12]
```diff
@@ -16,7 +16,6 @@ if(NOT CHECK_FOUND)
 endif()
 
 if(CHECK_FOUND)
-enable_testing()
```
Comment (abergeron) [inline gpuarray/test/CMakeLists.txt]: Why is this gone?

Comment (tsirif) [inline gpuarray/test/CMakeLists.txt]: It was duplicate. It was both in here and in <root>/CMakeLists.txt.

Comment (abergeron) [inline gpuarray/test/CMakeLists.txt]: It should be here and not in the root CMakeLists.txt

File: gpuarray/src/private_config.h (hunk 11 above)
Comment (abergeron) [inline gpuarray/src/private_config.h]: The comments on the `#else` and `#endif` are sometimes useful, but sometimes just annoying because the matching `#if` is just two lines above.  In the cases where there is significant separation between statements, a comment is useful, but in most cases you fixed, it is not.

File: gpuarray/include/gpuarray/buffer_blas.h (hunk 5 above)
Comment (tsirif) [inline gpuarray/include/gpuarray/buffer_blas.h]: I deleted some things that I wrote by mistake. I meant to write the following, sorry:  
"Google style advices for a header guard like this:
`#ifndef <project name>_<directories>_<file name>`. [...]"

**Comment by tsirif** [comment]
I removed the private_config.h and fixed /gpuarray/.gitignore, changed the names of implementation specific files and CMakeLists.txt accordingly.

File: gpuarray/include/gpuarray/buffer_blas.h (hunk 5 above)
Comment (abergeron) [inline gpuarray/include/gpuarray/buffer_blas.h]: It's not that much of a deal so ok.  It will just be annoying for future files.

File: gpuarray/src/private_config.h (hunk 11 above)
Comment (tsirif) [inline gpuarray/src/private_config.h]: Should I remove them wherever they are not needed? I will do it with a substitution script, it will be fast.

File: gpuarray/test/CMakeLists.txt (hunk 12 above)
Comment (tsirif) [inline gpuarray/test/CMakeLists.txt]: Hm for some reason cmake does not create a "test" make target when `enable_testing()` is included in a subdirectory's CMakeLists.txt.

Comment (abergeron) [inline gpuarray/test/CMakeLists.txt]: Ok then let's follow cmake's sillyness.

File: gpuarray/src/private_config.h (hunk 11 above)
Comment (abergeron) [inline gpuarray/src/private_config.h]: Yes if you have a script to do so.

**Comment by tsirif** [comment]
I think it is better now.

**Comment by nouiz** [comment]
Should we close this PR or there is good stuff that need to be finished in it?

**Comment by tsirif** [comment]
This PR was at the time ready to be merged in. But I do not know whether
you consider these changes to be good to your convenience  in terms of code
management and if there is need for it to be rebased on the current latest
commit to assure that nothing breaks.

If this style of directories is not preferable, of course you can close it
right away.

**Comment by nouiz** [comment]
@abergeron, it is your call.

**Comment by abergeron** [comment]
I'm not against a reorganization, but I would rather have an include folder at the top level where the public includes would move and remove the gpuarray_ prefix from the source files.

The proposed reorganisation would have made development painful by adding multiple levels of subfolders for little benefit (libgpuarray/gpuarray/src/gpuarray/*c).
